storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3574,6 +3589,294 @@ dberr_t fsp_sys_tablespace_validate()
 }
 #endif /* UNIV_DEBUG */
 
+/** Get the list of pages from the file-based list
+@param base        buffer block
+@param boffset     offset
+@param mtr         mini-transaction
+@param page_ranges store the pages exist in the list */
+static void flst_get_page_range(
+  const buf_block_t *base, uint16_t boffset, mtr_t *mtr,
+  range_set *page_ranges)
+{
+  const uint32_t len= flst_get_len(base->page.frame + boffset);
+  fil_addr_t addr= flst_get_first(base->page.frame + boffset);
+  for (uint32_t i= len; i--; )
+  {
+    ut_ad(addr.boffset >= FIL_PAGE_DATA);
+    ut_ad(addr.boffset < base->physical_size() - FIL_PAGE_DATA_END);
+    const buf_block_t *b=
+      buf_page_get_gen(page_id_t(base->page.id().space(), addr.page),
+                       base->zip_size(), RW_SX_LATCH, nullptr, BUF_GET, mtr);
+    ut_ad(b);
+
+    /* Descriptor offset to page number */
+    uint32_t page= addr.page * static_cast<uint32_t>(srv_page_size);
+    page += ((addr.boffset - XDES_ARR_OFFSET) / XDES_SIZE) * FSP_EXTENT_SIZE;
+    page_ranges->add_value_range(page, 64);
+    addr= flst_get_next_addr(b->page.frame + addr.boffset);
+    mtr->release_last_page();
+  }
+  ut_ad(addr.page == FIL_NULL);
+}
+
+/** Get the range of pages exist in the given segment
+@param header       segment header
+@param mtr          mini-transaction
+@param page_ranges  store the range of pages exist in this segment */
+static void fseg_get_pages(const fseg_header_t *header, mtr_t *mtr,
+                           range_set *page_ranges)
+{
+  buf_block_t *iblock;
+  fseg_inode_t *inode= fseg_inode_try_get(header, 0, 0, mtr, &iblock);
+  const uint16_t ioffset= uint16_t(inode - iblock->page.frame);
+  fseg_get_n_frag_pages(inode, page_ranges);
+
+  flst_get_page_range(
+    iblock, uint16_t(ioffset + FSEG_NOT_FULL), mtr, page_ranges);
+
+  flst_get_page_range(
+    iblock, uint16_t(ioffset + FSEG_FULL), mtr, page_ranges);
+
+  flst_get_page_range(
+    iblock, uint16_t(ioffset + FSEG_FREE), mtr, page_ranges);
+
+  return;
+}
+
+#ifdef UNIV_DEBUG
+/** Print the system tablespace page type information
+@return error code */
+static
+dberr_t fsp_system_print_pages()
+{
+  fprintf(stderr, "Print the pages in the system tablespace\n\n\t");
+  fil_space_t *space= fil_system.sys_space;
+  uint32_t n_words= 0;
+  dberr_t err= DB_SUCCESS;
+  mtr_t mtr;
+  mtr.start();
+  for (uint32_t i= 0; i < space->free_limit; i++)
+  {
+    if (n_words == 16)
+    {
+      fprintf(stderr, "\n\t");
+      n_words= 0;
+    }
```

Comment:
It could be good to display the page number in hexadecimal at the start of each line.